src/core/thread/mle_router.cpp

```diff
@@ -1438,7 +1438,11 @@ Error MleRouter::HandleAdvertisement(RxInfo &aRxInfo)
 
         if (routerCount > mRouterDowngradeThreshold && mRouterSelectionJitterTimeout == 0 &&
             HasMinDowngradeNeighborRouters() && HasSmallNumberOfChildren() &&
-            HasOneNeighborWithComparableConnectivity(route, routerId))
+            HasOneNeighborWithComparableConnectivity(route, routerId)
+#if OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE && OPENTHREAD_CONFIG_BORDER_ROUTER_REQUEST_ROUTER_ROLE
+            && !Get<NetworkData::Notifier>().IsEligibleForRouterRoleUpgradeAsBorderRouter()
```

Comment:
The check for 'IsEligibleForRouterRoleUpgradeAsBorderRouter()' will return false if the number of routers is equal to MAX_ROUTER.  In that particular case, then the BR will downgrade even though it's entitled to keep its state as Router.
In practice this condition will likely never occur. And perhaps it's even a good feature: e.g. in case of 2 BRs providing connectivity that are Router, and MAX_ROUTERS is reached, one of the two will downgrade to make some room for others to become Router if needed. So I would just leave it as is but good to be aware of.

Reply:
Yes. I agree. Thanks @EskoDijk.